text/0104-string-parameterization-output-spec.md

```diff
@@ -0,0 +1,264 @@
+
+Sentry performs parameterization in many different part of the application. The SDKs, Relay, and Ingest are three examples. There is no agreement between these parts beyond conventions. As a result, the output of parameterization is inconsistent.
+
+Here are three ways different SDKs might parameterize a SQL query:
+
+- `SELECT * FROM countries WHERE id = ?`
+- `SELECT * FROM countries WHERE id = %s`
+- `SELECT * FROM countries WHERE id = $1`
+
+As another example, the SDKs might pass the query `SELECT * FROM country WHERE code IN (?, ?)` to Ingest, which will further parameterize that as `SELECT * FROM country WHERE code IN (?)`. The string `(?, ?)` is technically only partially parameterized from the system's perspective.
+
+This has important downstream effects on user experience. A good parameterizer will correctly group similar strings together in a way that represents the original intent. If parameterization is successful, users will see groups that make sense to them. Good parameterization will also reduce cardinality, and therefore reduce data storage costs. A bad parameterizer will fill the users’s lists of transactions and spans with noisy similar descriptions that drown out the signal.
+
+This spec documents the desired output of a parameterizer. This makes it possible for different parameterizers to agree, conform to a standard, share code, and create consistency in the system.
+
+## Definitions
+
+- **Parameter**: A substring in a span description that represents a dynamic value that was substituted in by code. e.g., in the string `/users/17`, `17` is a likely parameter
+- **Parameterization**: The act of determining parameters, and replacing them with a substitution character. Also sometimes called "normalization"
+- **Parameterizer**: A piece of code, or a system that performs parameterization
+- **Substitution character**: A character or string of characters that denote a parameter. e.g., in the string `SELECT * FROM countries WHERE id = ?;` the `?` is a substitution character
+
+## Overall Guidelines
+
+- Parameterize *every* parameter. If a string contains multiple parameters, parameterize each one according to the relevant rules
+- Re-order unordered parameters in alphabetical order if it does not change the meaning of the string (e.g., see “Query Strings”)
```

Comment:
I'm not sure this is a good idea. I think the parameterized string should be as close to the original string as possible. If the parameterized string is later used to create a _hash_ or some kind of other unique identifier, that part of the code can decide to re-arrange things in order to ensure consistent hashing.

Reply:
What is the problem we're trying to solve with this point? Is the following reasoning fair?

1. SDKs capture parameters in a certain order, consistently. The order they capture the parameters is always the same.
2. We can consider the parameters to not be ordered, according to sorting rules we've considered "correct".
3. A set of parameters that is always in the same order is already ordered, by some rules that we may not know/have defined.
4. Thus, changing the order doesn't effectively make any improvements on the ordering, while introducing more complexity.

Reply:
> SDKs capture parameters in a certain order, consistently. The order they capture the parameters is always the same.

Is this always true?

The original goal was to make sure that different parameter order doesn't affect parameterization, in cases where the SDKs do _not_ capture parameters in a certain order consistently. Can SDKs guarantee that? If yes, we don't need this point. If not, we might want to specify that parameterization ensures the order

Reply:
I believe I wasn't clear enough. If SDKs don't modify the order in any way, then SDKs always extract it the same way. This means that the order is whatever users wrote, and it doesn't matter if it's alphabetically sorted or not. Does this make sense?